express_relay/sdk/js/src/index.ts

```diff
@@ -118,13 +120,157 @@ function checkTokenQty(token: { contract: string; amount: string }): TokenQty {
   };
 }
 
+type ClientOptions = FetchClientOptions & { baseUrl: string };
+
 export class Client {
-  private clientOptions?: ClientOptions;
+  public clientOptions: ClientOptions;
+  public websocket?: WebSocket;
+  public idCounter = 0;
+  public callbackRouter: Record<
+    string,
+    (response: components["schemas"]["ServerResultMessage"]) => void
+  > = {};
+  private websocketOpportunityCallback?: (
+    opportunity: Opportunity
+  ) => Promise<void>;
 
-  constructor(clientOptions?: ClientOptions) {
+  constructor(clientOptions: ClientOptions) {
     this.clientOptions = clientOptions;
   }
 
+  private connectWebsocket() {
+    const websocketEndpoint = new URL(this.clientOptions.baseUrl);
+    websocketEndpoint.protocol =
+      websocketEndpoint.protocol === "https:" ? "wss:" : "ws:";
+    websocketEndpoint.pathname = "/v1/ws";
+
+    this.websocket = new WebSocket(websocketEndpoint.toString());
+    this.websocket.on("message", async (data) => {
+      const message:
+        | components["schemas"]["ServerResultResponse"]
+        | components["schemas"]["ServerUpdateResponse"] = JSON.parse(
+        data.toString()
+      );
+      if ("id" in message && message.id) {
+        const callback = this.callbackRouter[message.id];
+        if (callback !== undefined) {
+          callback(message);
+          delete this.callbackRouter[message.id];
+        }
+      } else if ("type" in message && message.type === "new_opportunity") {
+        if (this.websocketOpportunityCallback !== undefined) {
+          const convertedOpportunity = this.convertOpportunity(
+            message.opportunity
+          );
+          if (convertedOpportunity !== undefined) {
+            await this.websocketOpportunityCallback(convertedOpportunity);
```

Comment:
What happens if this is undefined? Maybe it be nice to take this lambda function in constructor (if it's critical for the client)

Reply:
They will see the logs about this. I would rather improve this when we get some feedbacks from searchers